kafkabp/consumer.go

```diff
@@ -0,0 +1,222 @@
+package kafkabp
+
+import (
+	"context"
+	"io"
+	"sync"
+	"sync/atomic"
+
+	"github.com/Shopify/sarama"
+
+	"github.com/reddit/baseplate.go/metricsbp"
+	"github.com/reddit/baseplate.go/tracing"
+)
+
+// ConsumeMessageFunc is a function type for consuming consumer messages.
+type ConsumeMessageFunc func(ctx context.Context, msg *sarama.ConsumerMessage) error
+
+// ConsumeErrorFunc is a function type for consuming consumer errors.
+type ConsumeErrorFunc func(err error)
+
+// consumer is an instance of a Kafka consumer.
+type consumer struct {
+	cfg ConsumerConfig
+	sc  *sarama.Config
+
+	consumer           atomic.Value // sarama.Consumer
+	partitions         atomic.Value // []int32
+	partitionConsumers atomic.Value // []sarama.PartitionConsumer
+
+	closed          int64
+	consumeReturned int64
+	offset          int64
+
+	wg sync.WaitGroup
+}
+
+// Consumer defines the interface of a consumer struct.
+type Consumer interface {
+	io.Closer
+
+	Consume(ConsumeMessageFunc, ConsumeErrorFunc) error
+
+	// IsHealthy returns false after Consume returns.
+	IsHealthy() bool
+}
+
+// NewConsumer creates a new Kafka consumer. Unlike a group consumer (which
+// delivers every message exactly once by having one ClientID assigned to every
+// consumer in the group), this consumer is used for consuming some
+// configuration or data by all running consumer instances. This is why the
+// ClientID provided to NewConsumer's ConsumerConfig must be unique.
+func NewConsumer(cfg ConsumerConfig) (Consumer, error) {
+	sc, err := cfg.NewSaramaConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	kc := &consumer{
+		cfg:    cfg,
+		sc:     sc,
+		offset: sc.Consumer.Offsets.Initial,
+	}
+
+	// Initialize Sarama consumer and set atomic values.
+	if err := kc.reset(); err != nil {
+		return nil, err
+	}
+
+	return kc, nil
+}
+
+func (kc *consumer) getConsumer() sarama.Consumer {
+	c, _ := kc.consumer.Load().(sarama.Consumer)
+	return c
+}
+
+func (kc *consumer) getPartitions() []int32 {
+	p, _ := kc.partitions.Load().([]int32)
+	return p
+}
+
+func (kc *consumer) getPartitionConsumers() []sarama.PartitionConsumer {
+	pc, _ := kc.partitionConsumers.Load().([]sarama.PartitionConsumer)
+	return pc
+}
+
+// reset recreates the consumer and assigns partitions.
+func (kc *consumer) reset() error {
+	if c := kc.getConsumer(); c != nil {
+		if err := c.Close(); err != nil {
+			return err
```

Comment:
I don't think this is the right way to do it. Close errors are usually considered much less severe compare to other errors (especially close errors from readers), and they can usually just be logged and ignored. If we just return the error here that means close errors will cause the whole resetting and consuming loop to break and fail.

But in prod I think this rarely happens so maybe this is ok? I still prefer this to be just logged and ignored instead.

Reply:
I figure adding a `log.Wrapper` to the Config and letting the user decide if they want to set it is the right approach. Thanks for the guidance!